### PR TITLE
internal: fix GO_AWAY deadlock

### PR DESCRIPTION
**Pull request by jadekler**
A deadlock can occur when a GO_AWAY is followed by a connection closure. This
happens because onClose needlessly closes the current ac.transport: if a
GO_AWAY already occured, and the transport was already reset, then the later
closure (of the original address) sets ac.transport - which is now healthy -
to nil.

The manifestation of this problem is that picker_wrapper spins forever trying
to use a READY connection whose ac.transport is nil.

### CODE REVIEW

File: test/end2end_test.go [hunk 1]
```diff
@@ -6994,3 +6995,96 @@ func testLargeTimeout(t *testing.T, e env) {
...
+			// called.
+			client.UnaryCall(ctx, &testpb.SimpleRequest{})
+		}
+	}
+
+	// Close connection 1.
+	s1.Stop()
+
+	// Wait for client to close.
+	time.Sleep(100 * time.Millisecond)
```
Comment (dfawley) [inline test/end2end_test.go]: Would this work instead? `<-stream.Context().Done()` ?

File: test/end2end_test.go [hunk 2]
```diff
@@ -6994,3 +6995,99 @@ func testLargeTimeout(t *testing.T, e env) {
...
+		},
+	}
+	testpb.RegisterTestServiceServer(s1, ts1)
+	go s1.Serve(lis1)
+
+	lis2, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis2.Close()
```
Comment (dfawley) [inline test/end2end_test.go]: `s2.Stop()` will take care of this.  No need to do it separately.

File: test/end2end_test.go [hunk 3]
```diff
@@ -6994,3 +6995,99 @@ func testLargeTimeout(t *testing.T, e env) {
...
+func TestGoAwayThenClose(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	ctx := context.Background()
+
+	lis1, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis1.Close()
```
Comment (dfawley) [inline test/end2end_test.go]: `s1.Stop()` will take care of this.  No need to do it separately.

File: test/end2end_test.go [hunk 4]
```diff
@@ -6994,3 +6995,99 @@ func testLargeTimeout(t *testing.T, e env) {
...
 	}
 }
+
+// Proxies typically send GO_AWAY followed by connection closure a minute or so later. This
+// test ensures that the connection is re-created after GO_AWAY and not affected by the
+// subsequent (old) connection closure.
+func TestGoAwayThenClose(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	ctx := context.Background()
```
Comment (dfawley) [inline test/end2end_test.go]: Avoid infinite delays:

`context.WithTimeout(context.Background(), 20*time.Second)`?

File: test/end2end_test.go (hunk 1 above)
Comment (jadekler) [reply test/end2end_test.go]: By stream are you suggesting, `stream, err := client.FullDuplexCall(ctx)`?

If so, it does not appear to ever finish.

File: test/end2end_test.go (hunk 3 above)
Comment (jadekler) [reply test/end2end_test.go]: Done

File: test/end2end_test.go (hunk 2 above)
Comment (jadekler) [reply test/end2end_test.go]: Done

File: test/end2end_test.go (hunk 4 above)
Comment (jadekler) [reply test/end2end_test.go]: Done

**Comment by jadekler** [comment]
PTAL

File: test/end2end_test.go [hunk 5]
```diff
@@ -7055,7 +7056,8 @@ func TestGoAwayThenClose(t *testing.T) {
 
 	client := testpb.NewTestServiceClient(cc)
 
-	// Should go on connection 1.
+	// Should go on connection 1. We use a long-lived RPC because it will cause GracefulStop to send GO_AWAY, but the
+	// connection doesn't get closed until the server stops and the client receives.
 	stream, err := client.FullDuplexCall(ctx)
 	if err != nil {
 		t.Fatalf("UnaryCall(_) = _, %v; want _, nil", err)
```
Comment (dfawley) [inline test/end2end_test.go]: *FullDuplexCall

**Pull request merged**